src/backtracking/hamiltonian_cycle.rs

```diff
@@ -0,0 +1,298 @@
+//! This module provides functionality to find a Hamiltonian cycle in a directed or undirected graph.
+//! Source: [Wikipedia](https://en.wikipedia.org/wiki/Hamiltonian_path_problem)
+
+use std::collections::HashSet;
+
+/// Represents potential errors when finding hamiltonian cycle on an adjacency matrix.
+#[derive(Debug, PartialEq, Eq)]
+pub enum FindHamiltonianCycleError {
+    /// Indicates that the adjacency matrix is empty.
+    EmptyAdjMat,
+    /// Indicates that the adjacency matrix is not square.
+    ImproperAdjMat,
+    /// Indicates that the starting vertex is out of bounds.
+    StartOutOfBound,
+}
+
+/// Represents a graph using an adjacency matrix.
+struct Graph {
+    /// The adjacency matrix representing the graph.
+    adjacency_matrix: Vec<Vec<bool>>,
+}
+
+impl Graph {
+    /// Creates a new graph with the provided adjacency matrix.
+    ///
+    /// # Arguments
+    ///
+    /// * `adjacency_matrix` - A square matrix where each element indicates
+    ///                        the presence (`true`) or absence (`false`) of an edge
+    ///                        between two vertices.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the graph if successful, or an `FindHamiltonianCycleError` if there is an issue with the matrix.
+    fn new(adjacency_matrix: Vec<Vec<bool>>) -> Result<Self, FindHamiltonianCycleError> {
+        // Check if the adjacency matrix is empty.
+        if adjacency_matrix.is_empty() {
+            return Err(FindHamiltonianCycleError::EmptyAdjMat);
+        }
+
+        // Validate that the adjacency matrix is square.
+        if adjacency_matrix
+            .iter()
+            .any(|row| row.len() != adjacency_matrix.len())
+        {
+            return Err(FindHamiltonianCycleError::ImproperAdjMat);
+        }
+
+        Ok(Self { adjacency_matrix })
+    }
+
+    /// Returns the number of vertices in the graph.
+    fn num_vertices(&self) -> usize {
+        self.adjacency_matrix.len()
+    }
+
+    /// Determines if it is safe to include vertex `v` in the Hamiltonian cycle path.
+    ///
+    /// # Arguments
+    ///
+    /// * `v` - The index of the vertex being considered.
+    /// * `visited` - A reference to the set of visited vertices.
+    /// * `path` - A reference to the current path being explored.
+    /// * `pos` - The position of the current vertex being considered.
+    ///
+    /// # Returns
+    ///
+    /// `true` if it is safe to include `v` in the path, `false` otherwise.
+    fn is_safe(&self, v: usize, visited: &HashSet<usize>, path: &[usize], pos: usize) -> bool {
```

Comment:
Using `HashSet` here might be an overkill. Since the vertices are numbered from zero, one could consider using `Vec<bool>`, with some _possible_ performance gain.

Reply:
It's a good option to implement. Using a `Vec<bool>` instead of a `HashSet<usize>` can indeed provide a performance boost since it allows for constant-time access and updates, and it typically consumes less memory.